---
title: "Cell Lines"
subtitle: "Simplification of a simple scRNA-seq dataset of 5 cancer cell lines from (Tian et al., 2019)[https://doi.org/10.1038/s41592-019-0425-8]"
author: "Mariia Bilous, Aurelie Gabriel, Leonard Herault and David Gfeller"
output:
  md_document:
    toc: true # table of content true



---

# Workbook structure

We first run a **standard scRNA-seq data analysis pipeline** (i.e., data normalization, feature selection, dimensionality reduction, visualization, clustering and differential expression analysis) using the [Seurat](https://satijalab.org/seurat/index.html) framework. Then, we **simplify** the same dataset by computing *metacells* (i.e., grouping transcriptionally highly similar single cells into metacells). For this, we will use a method developed in our group called [SuperCell](https://github.com/GfellerLab/SuperCell). We then perform the same **standard downstream analysis** of *metacells* and compare the results to those obtained at the single-cell level. Additionally, we provide code to run sample-weighted downstream analysis that accounts for the metacells size. 


We also provide some scripts to build *metacells* using alternative approaches including  [Metacell-2](https://metacells.readthedocs.io/en/latest/readme.html) ([script to build metacells with Metacell-2](https://github.com/GfellerLab/SIB_workshop/blob/main/workbooks/Metacell2.ipynb)) and [SEACell](https://github.com/dpeerlab/SEACells) ([script to build metacells with SEACells](https://github.com/GfellerLab/SIB_workshop/blob/main/workbooks/SEACells.ipynb)). To avoid any issues related to the installation of those packages or data transfrering from Python to R, we also privide precomputed results of metacell partition using those two methods (see \ref{Metacell2} and \ref{SEACells})


```{r Load libraries, message=FALSE, warning=FALSE}
# make a data library (cell lines or Zilionis)
library(SuperCell)
library(Seurat)
library(dplyr)
```

```{r Load data}
proj.name    <- 'cell_lines'
.color.cell.type <- c("A549" = "#E69F00", "H838" = "#56B4E9", "H1975" = "#009E73", "H2228" = "#F0E442", "HCC827" = "#CC79A7")
data.folder  <- file.path("..", "data", proj.name)

# load single-cell (sc) count matrix and cell metadata 
sc.counts <- readRDS(file.path(data.folder, "sc_counts_filtered.Rds"))
sc.meta   <- readRDS(file.path(data.folder, "sc_meta_filtered.Rds"))

# Make sure metadata and count matrix have the same cells in the same order
if(!identical(rownames(sc.meta), colnames(sc.counts))){
  stop("Metadata (`sc.meta`) does not correspond to the count matrix (`sc.counts`)")
}
```

# Single-cell level
## Standard downstream analysis  
Run a brief analysis at the single-cell level, lets's use the common [Seurat](https://satijalab.org/seurat/index.html) pipeline.

Depending on the participants' experience with Seurat, we can either separately run these steps of the pre-processing or 'skip it' with 'one-line' command. For the moment, it is more step-by -step pre-processing, that can be replaced with : 
`sc <-  NormalizeData(sc, verbose=FALSE) %>% FindVariableFeatures(selection.method = "disp", nfeatures = 1000, verbose=FALSE) %>% ScaleData(verbose=FALSE) %>% RunPCA(verbose=FALSE)`

```{r}
set.seed(12345)
sc <- CreateSeuratObject(counts = sc.counts, project = proj.name, meta.data = sc.meta)
sc
```

### Pre-processing
#### Data normalization
```{r Normalize data and compute a set of highly variable genes, message=FALSE, warning=FALSE}
sc <- NormalizeData(sc, verbose=FALSE)

sc <- FindVariableFeatures(
  sc, 
  selection.method = "disp", # "vst" is default
  nfeatures = 1000,
  verbose=FALSE
  )

hvg <- VariableFeatures(sc, verbose=FALSE)

# Plot variable features 
plot1 <- VariableFeaturePlot(sc)
LabelPoints(plot = plot1, points = hvg[1:20], repel = TRUE)
```

#### Scaling and dimensionality redution
```{r message=FALSE, warning=FALSE}
sc <- ScaleData(sc, verbose=FALSE)
sc <- RunPCA(sc, verbose=FALSE)

# Plot PCA (2D representation of scRNA-seq data) colored by cell line
DimPlot(sc, reduction = "pca", group.by = "cell_line", cols = .color.cell.type)
```

### UMAP (non-linear dimensionality reduction)
```{r UMAP, message=FALSE, warning=FALSE}
sc <- RunUMAP(sc,  dims = 1:10)

# Plot UMAP (2D representation of scRNA-seq data) colored by cell line
DimPlot(sc, reduction = "umap", group.by = "cell_line", cols = .color.cell.type)
```

### Clustering
```{r message=FALSE, warning=FALSE}

sc <- FindNeighbors(sc, dims = 1:10)
sc <- FindClusters(sc, resolution = 0.05)

# As it is a toy example with well defined cell types (i.e., cell lines), unsupervised clustering fully recapitulates cell line annotation 
table(sc@active.ident, sc$cell_line)
DimPlot(sc, reduction = "umap", group.by = "ident")

```

### Differential expression analysis (!? TO remove this part if COVID dataset contains such analysis)
#### Find Markers of cell lines
```{r message=FALSE, warning=FALSE}
# Set idents to cell lines (as clusters are the same as cell lines)
Idents(sc) <- "cell_line" 
levels(sc) <- sort(levels(sc))

# Compute upregulated genes in each cell line (versus other cells)
sc.all.markers <-  FindAllMarkers(sc, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25, test.use = "t")
saveRDS(sc.all.markers, file = file.path(data.folder, "output", "sc_all_markers.Rds"))

# Load markers (backup)
#sc.all.markers <- readRDS(file = file.path(data.folder, "output", "sc_all_markers.Rds"))

# Top markers (select top markers of each cell line)
sc.top.markers <- sc.all.markers %>%
   group_by(cluster) %>%
    slice_max(n = 2, order_by = avg_log2FC)

sc.top.markers
```

#### Plot the expression of some found markers 
```{r fig.height=5, fig.width=9}
VlnPlot(sc, features = sc.top.markers$gene[c(seq(1, 9, 2), seq(2, 10, 2))], ncol = 5, pt.size = 0.0, cols = .color.cell.type)
```


# Data simplification (coarse-graining) -- Construction of *metacells*

Here we compute metacells using our method called [SuperCell](https://github.com/GfellerLab/SuperCell), but equally, metacells can be computed with [Metacell](https://github.com/tanaylab/metacell), [Metacell-2](https://metacells.readthedocs.io/en/latest/readme.html) or [SEACell](https://github.com/dpeerlab/SEACells) algorithms and we will see some examples below (see \ref{Metacell2} and \ref{SEACells}).
```{r Set parameters}
gamma <- 20 # Graining level

# Compute metacells using SuperCell package
MC <- SCimplify(
  X = GetAssayData(sc), # single-cell log-normalized gene expression data
  genes.use = hvg, 
  gamma = gamma,
  n.pc = 10
)

# Compute gene expression of metacells by simply averaging gene expression within each metacell
MC.ge <- supercell_GE(
  ge = GetAssayData(sc),
  groups = MC$membership
)

# Alternatively, counts can be averaged (summed up) followed by a lognormalization step (this approach is used in the MetaCell and SEACell algorithms)
if(0){
  MC.counts <- supercell_GE(
    ge = GetAssayData(sc, slot = "counts"),
    mode = "sum", # summing counts instead of the default averaging
    groups = MC$membership
  )
  
  MC.ge <- Seurat::LogNormalize(MC.counts, verbose = FALSE)
}
```

## Analysis of *metacells*
There are two options to perform the downstream analysis: 

* **standard** when we treat metacells as single-cell and apply a standard pipeline
* **sample-weighted** when we account for a metacell size at each step of the analysis 

### Pre-processing

#### Transfer metadata (annotate metacell to a certain cell line)
Since the cell line information is available in this dataset, we can annotate metacells to a certain cell line. Each metacell is annotated to the most abundant cell type in it.
This also allows us to compute metacell *purity*, that is defined as a proportion of the most abundant cell type (use `method = "max_proportion"`) or as Shannon entropy (use `method = "entropy"`).

```{r metacell annotation to cell line and their purity}
# Annotate metacells to cells line
MC$cell_line <- supercell_assign(
  cluster = sc.meta$cell_line,          # single-cell assignment to cell lines 
  supercell_membership = MC$membership,  # single-cell assignment to metacells
  method = "absolute" # available methods are c("jaccard", "relative", "absolute"), function's help() for explanation
)

# Compute purity of metacells as :
#  * a proportion of the most abundant cell type withing metacells (`method = `"max_proportion)
#  * an entropy of cell type within metacells (`method = "entropy"`)
method_purity <- c("max_proportion", "entropy")[1]
MC$purity <- supercell_purity(
  clusters = sc.meta$cell_line,
  supercell_membership = MC$membership, 
  method = method_purity
)

# Metacell purity distribution
summary(MC$purity)
#hist(MC$purity, main = paste0("Purity of metacells \nin terms of cell line composition (", method_purity,")"))
```

#### Visualize data using  metacell network 
```{r metacell plot metacell network color cell line}
supercell_plot(
  MC$graph.supercells, 
  group = MC$cell_line, 
  color.use = .color.cell.type,
  seed = 1, 
  alpha = -pi/2,
  main  = "Metacells colored by cell line assignment"
)
```

## Standard downstream analysis of *metacells*

For the standard downstream analysis, we can use the well-established [Seurat](https://satijalab.org/seurat/index.html) pipeline

### Create Seurat object to perform standard downstream analysis 
When creating Seurat Object, we perform sample-weighted scaling of gene expression data and sample-weighted PCA (with the weigh being the metacell size).

```{r}
MC.seurat <- supercell_2_Seurat(
  SC.GE = MC.ge, 
  SC = MC, 
  fields = c("cell_line", "purity"),
  var.genes = MC$genes.use,
  N.comp = 10
)

Idents(MC.seurat) <- "cell_line"

MC.seurat <- RunUMAP(MC.seurat, dims = 1:10)

DimPlot(MC.seurat, cols = .color.cell.type, reduction = "umap")
```

### Clustering
Seurat clustering
```{r}
MC.seurat <- FindClusters(MC.seurat, resolution = 0.5)
DimPlot(MC.seurat, reduction = "umap")

```

### DEA of cell lines in *metacells*
#### Find Markers of cell lines
```{r message=FALSE, warning=FALSE}
# Set idents to cell lines (as clusters are the same as cell lines)
Idents(MC.seurat) <- "cell_line"
levels(MC.seurat) <- sort(levels(Idents(MC.seurat)))

# Compute upregulated genes in each cell line (versus other cells)
MC.seurat.all.markers <-  FindAllMarkers(
  MC.seurat, 
  only.pos = TRUE,
  min.pct = 0.25, 
  logfc.threshold = 0.25, 
  test.use = "t"
)
saveRDS(MC.seurat.all.markers, file = file.path(data.folder, "output", paste0("MC_gamma_", gamma, "_all_markers_seurat.Rds")))

# Load markers (backup)
#MC.seurat.all.markers <- readRDS(file = file.path(data.folder, "output", "MC_gamma_20_all_markers_seurat.Rds"))

# Top markers (select top markers of each cell line)
MC.seurat.top.markers <- MC.seurat.all.markers %>%
   group_by(cluster) %>%
    slice_max(n = 2, order_by = avg_log2FC)

MC.seurat.top.markers
```

#### Plot the expression of some found markers 
```{r fig.height=5, fig.width=9}
genes.to.plot <- MC.seurat.top.markers$gene[c(seq(1, 9, 2), seq(2, 10, 2))]
VlnPlot(MC.seurat, features = genes.to.plot, ncol = 5, pt.size = 0.0, cols = .color.cell.type)
```

### Plot gene-gene correlation at single-cell and metacell levels (! TO find better examples)
```{r}
gene_x <- MC$genes.use[500:505] #500
gene_y <- MC$genes.use[550:555] #600

alpha <- 0.7

p.SC <- supercell_GeneGenePlot(GetAssayData(sc, slot = "data"), gene_x = gene_x, gene_y = gene_y, clusters = sc$cell_line, color.use = .color.cell.type, sort.by.corr = F, alpha = alpha)
p.SC$p

p.MC <- supercell_GeneGenePlot(MC.ge, gene_x = gene_x, gene_y = gene_y, supercell_size = MC$supercell_size, clusters = MC$cell_line, color.use = .color.cell.type, sort.by.corr = F, alpha = alpha)
p.MC$p

```

## Alternative or Sample-weighted downstream analysis of *metacells*

For the sample-weighted analysis, we use a pipeline avalable with our [SuperCell](https://github.com/GfellerLab/SuperCell) package.

### Dimensionality reduction 
```{r metacell PCA}
MC$PCA <- supercell_prcomp(
  Matrix::t(MC.ge),
  genes.use = MC$genes.use,  # or a new set of HVG can be computed
  supercell_size = MC$supercell_size, # provide this parameter to run sample-weighted version of PCA,
  k = 10
)

MC$UMAP <- supercell_UMAP(
  SC = MC,
  PCA_name = "PCA",
  n_neighbors = 50 # large number to repel cells 
)

supercell_DimPlot(
  MC, 
  groups = MC$cell_line,
  dim.name = "UMAP", 
  title = paste0("UMAP of metacells colored by cell line assignment"),
  color.use = .color.cell.type
)

```

### Clustering (sample-weighted hclust)

Sample-weighted clustering computed with the hierarchical clustering, that may accounts for sample weights
```{r metacell clustering}
# compute distance among metacells
D                <- dist(MC$PCA$x)

# cluster metacells
MC$clustering    <- supercell_cluster(D = D, k = 5, supercell_size = MC$supercell_size)

# Plot clustering result
supercell_DimPlot(
  MC, 
  groups = factor(MC$clustering$clustering),
  dim.name = "UMAP", 
  title = paste0("UMAP of metacells colored by metacell clustering")
)
```

```{r Consistency between clustering and cell line assignment }
table(MC$cell_line, MC$clustering$clustering)
```

### Differential expression analysis in metacells (sample-weighted t-test)
```{r DEA metacells}
# Compute upregulated genes in each cell line (versus other cells)
MC.all.markers <- supercell_FindAllMarkers(
  ge = MC.ge, 
  clusters = MC$cell_line, 
  supercell_size = MC$supercell_size,
  only.pos = TRUE, 
  min.pct = 0.25, 
  logfc.threshold = 0.25
)

saveRDS(MC.all.markers, file = file.path(data.folder, "output",  paste0("MC_gamma_", gamma, "_all_markers.Rds")))

# Load markers (backup)
#MC.all.markers <- readRDS(file = file.path(data.folder, "output", "paste0("MC_gamma_", gamma, "_all_markers.Rds")))

# Transform the ourput of `supercell_FindAllMarkers()` to be in the format of the `Seurat::FindAllMarkers()`
MC.all.markers.df <- data.frame()
for(cl in names(MC.all.markers)){
  cur <- MC.all.markers[[cl]]
  cur$cluster <- cl
  cur$gene <- rownames(cur)
  cur$avg_log2FC <- cur$logFC
  MC.all.markers.df <- rbind(MC.all.markers.df, cur)
}

# Top markers (select top markers of each cell line)
MC.top.markers <- MC.all.markers.df %>%
   group_by(cluster) %>%
    slice_max(n = 2, order_by = avg_log2FC)

```

#### Plot the expression of some found markers (in metacells)
```{r fig.height=5, fig.width=9}

supercell_VlnPlot(
  ge = MC.ge, 
  supercell_size = MC$supercell_size, 
  clusters = MC$cell_line,
  features = MC.top.markers$gene[c(seq(1, 9, 2), seq(2, 10, 2))],
  color.use = .color.cell.type,
  ncol = 5)

```


# Alternative constructions of metacells

## Metacell construction with [Metacell-2](https://metacells.readthedocs.io/en/latest/Metacells_Vignette.html)

Metacell concept is not limited to the SuperCell algorithm and metacells can be computed using the [Metacell-2](https://metacells.readthedocs.io/en/latest/Metacells_Vignette.html) and [SEACells](https://github.com/dpeerlab/SEACells) algorithms. Since both of the methods are implemented in Python, we provide scripts to buil metacells of the same cell line dataset either with [Metacell-2](https://github.com/GfellerLab/SIB_workshop/blob/main/workbooks/SEACells.ipynb) or with [SEACells](https://github.com/GfellerLab/SIB_workshop/blob/main/workbooks/Metacell2.ipynb). In the following sections you can find 2 examples how to use the output of those algorithms to obtain SuperCell-like output, which you can use for the downstream analysis as was demonstrated above. 

Load pre-computed metacells with the Metacell-2 approach. See this [workbook](https://github.com/GfellerLab/SIB_workshop/blob/main/workbooks/Metacell2.ipynb) to reproduce the results or to run your own Metacell-2 construction.

**Note**, that to read python output, you will need the [anndata](https://cran.r-project.org/web/packages/anndata/index.html) library that may need some effort to install correctly. In case you faced any difficulties reading .h5ad anndata object (file `"seacells_gamma_20.h5ad"`), you can load the converted to the SuperCell-like .Rds object (file `"seacells_gamma_20.Rdata"`) or Seurat object (file `"seacells_gamma_20_seurat.Rds"`)
```{r Metacell2}
I_HAVE_ANNDATA_PACKAGE <- TRUE
fname <- "metacell2_gamma_20"

if(I_HAVE_ANNDATA_PACKAGE){
  library(anndata)
  ## Load pre-computed metacell partition obtained with Metacell-2 (anndata)
  
  ## comment this and provide R object directly in case anndata will rise errors 
  metacell2_adata  <- read_h5ad(file.path(data.folder, "output", paste0(fname, ".h5ad"))) # Metacell as anndata object
  obs.sc           <- metacell2_adata$uns$sc.obs # obs (cell meta data) of single-cell data used to build metacells with Metacell2
  
  metacell2        <- anndata_2_supercell(metacell2_adata) 
  
  ## Annotate metacells to cell lines
  metacell2$SC_cell_line <- supercell_assign(
    sc.meta$cell_line, 
    supercell_membership = metacell2$membership
  )
  
  ## Convert to Seurat object
  metacell2.seurat <- supercell_2_Seurat(
    SC.GE = metacell2$SC.counts, 
    is.log.normalized = FALSE, # Note, that we provided count matrix in contrast to the SuperCell output
    SC = metacell2, 
    fields = c("SC_cell_line", "supercell_size")
  )
  
  ## Add metadata that we stored 
  metacell2.seurat <- AddMetaData(metacell2.seurat, metadata = metacell2$SC.meta)
  Idents(metacell2.seurat) <- "SC_cell_line"
  levels(metacell2.seurat) <- sort(levels(metacell2.seurat))
  
  saveRDS(metacell2, file = file.path(data.folder, "output", paste0(fname, ".Rds")))
  saveRDS(metacell2.seurat, file = file.path(data.folder, "output", paste0(fname, "_seurat.Rds")))
  
} else {
  metacell2        <- readRDS(file = file.path(data.folder, "output", paste0(fname, ".Rds")))
  metacell2.seurat <- readRDS(file = file.path(data.folder, "output", paste0(fname, "_seurat.Rds")))
}
```
## Metacell construction with [SEACells](https://github.com/dpeerlab/SEACells)

Load pre-computed metacells with the SEACells approach. See this [workbook](https://github.com/GfellerLab/SIB_workshop/blob/main/workbooks/SEACells.ipynb) to reproduce the results or to run your own SEACells construction.
```{r SEACells}
I_HAVE_ANNDATA_PACKAGE <- TRUE
fname <- "seacells_gamma_20"

if(I_HAVE_ANNDATA_PACKAGE){
  library(anndata)
  ## Load pre-computed metacell partition obtained with SEACells (anndata)
  
  ## comment this and provide R object directly in case anndata will rise errors 
  seacell_adata    <- read_h5ad(file.path(data.folder, "output", paste0(fname, ".h5ad"))) # SEACells as anndata object
  obs.sc           <- seacell_adata$uns$sc.obs # obs (cell meta data) of single-cell data used to build metacells with SEACells
  
  seacell          <- anndata_2_supercell(seacell_adata) 
  
  ## Annotate metacells to cell lines
  seacell$SC_cell_line <- supercell_assign(
    sc.meta$cell_line, 
    supercell_membership = seacell$membership
  )
  
  ## Convert to Seurat object
  seacell.seurat <- supercell_2_Seurat(
    SC.GE = seacell$SC.counts, 
    is.log.normalized = FALSE, # Note, that we provided count matrix in contrast to the SuperCell output
    SC = seacell, 
    fields = c("SC_cell_line", "supercell_size")
  )
  
  Idents(seacell.seurat) <- "SC_cell_line"
  levels(seacell.seurat) <- sort(levels(seacell.seurat))
  
  saveRDS(seacell, file = file.path(data.folder, "output", paste0(fname, ".Rds")))
  saveRDS(seacell.seurat, file = file.path(data.folder, "output", paste0(fname, "_seurat.Rds")))
  
} else {
  seacell        <- readRDS(file = file.path(data.folder, "output", paste0(fname, ".Rds")))
  seacell.seurat <- readRDS(file = file.path(data.folder, "output", paste0(fname, "_seurat.Rds")))
}
```


```{r Sessioninfo, echo=FALSE}
sessioninfo::session_info()%>%
  details::details(
    summary = 'Current session info',
    open    = TRUE
  )
```


